rust/pattern/expression/operation.rs

```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    fn precedence(&self) -> Precedence {
+        match self {
+            Operation::Addition { .. } | Operation::Subtraction { .. } => Precedence::Addition,
+            Operation::Multiplication { .. } | Operation::Division { .. } | Operation::Modulo { .. } => {
+                Precedence::Multiplication
+            }
+            Operation::Power { .. } => Precedence::Exponentiation,
+        }
+    }
+
+    fn associativity(&self) -> Associativity {
+        match self {
+            Operation::Addition { .. } | Operation::Multiplication { .. } => Associativity::Associative,
+            Operation::Subtraction { .. } | Operation::Division { .. } | Operation::Modulo { .. } => {
+                Associativity::Left
+            }
+            Operation::Power { .. } => Associativity::Right,
+        }
+    }
+
+    fn arity(&self) -> Arity {
+        Arity::Binary
+    }
+
+    pub fn references_recursive(&self) -> Box<dyn Iterator<Item = &Reference> + '_> {
+        match self.arity() {
+            Arity::Binary => Box::new(
+                self.left().unwrap().references_recursive().chain(self.right().unwrap().references_recursive()),
+            ),
+        }
+    }
+}
+
+impl SubExpression for Operation {}
+
+impl fmt::Display for Operation {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.arity() {
+            Arity::Binary => {
+                let mut left_operand = format!("{}", self.left().unwrap());
+                if let Some(Expression::Operation(left_operation)) = self.left() {
+                    if let Associativity::Right = self.associativity() {
+                        if left_operation.precedence() == self.precedence() {
+                            left_operand = format!("( {left_operand} )");
+                        }
+                    }
+                    if left_operation.precedence() > self.precedence() {
+                        left_operand = format!("( {left_operand} )");
+                    }
+                }
+                let mut right_operand = format!("{}", self.right().unwrap());
+                if let Some(Expression::Operation(right_operation)) = self.right() {
+                    if let Associativity::Left = self.associativity() {
+                        if right_operation.precedence() == self.precedence() {
+                            right_operand = format!("( {right_operand} )");
+                        }
+                    }
+                    if right_operation.precedence() > self.precedence() {
+                        right_operand = format!("( {right_operand} )");
+                    }
+                }
+                write!(f, "{} {} {}", left_operand, self.op_token(), right_operand)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Precedence {
+    Addition,
+    Multiplication,
+    Exponentiation,
+}
+
+impl Precedence {
+    fn discriminant(&self) -> usize {
```

Comment:
can we inline this into the `enum` directly? cc @dmitrii-ubskii

Reply:
If we remove this method, we need to repeat this match twice in `Ord`. And for me it's nice to have explicit levels of precedence.

Reply:
It is a standard practice for attaching constants to `enum` variants.